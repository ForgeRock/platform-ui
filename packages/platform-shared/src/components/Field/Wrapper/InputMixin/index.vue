<!-- Copyright 2020 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<script>
import {
  delay,
  noop,
} from 'lodash';

export default {
  name: 'InputMixin',
  props: {
    /**
     * if field should be disabled.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Input name.
     */
    fieldName: {
      type: String,
      default: '',
    },
    /**
     * Related text that displays underneath field.
     */
    helpText: {
      type: String,
      default: '',
    },
    /**
     * Boolean to render label and help text as html.
     */
    isHtml: {
      type: Boolean,
      default: false,
    },
    /**
     * Placeholder value.
     */
    label: {
      type: String,
      default: '',
    },
    /**
     * Function used to validate data.
     */
    validator: {
      type: Function,
      default: () => noop,
    },
    /**
     * v-model value
     */
    value: {
      type: [Array, Object, Number, String, Boolean],
      default: '',
    },
  },
  data() {
    return {
      errorMessages: [],
      floatLabels: false,
      hideLabel: true,
      id: null,
      inputValue: '',
    };
  },

  beforeMount() {
    // eslint-disable-next-line no-underscore-dangle
    this.id = `floatingLabelInput${this._uid}`;
  },
  mounted() {
    delay(() => {
      if (navigator.userAgent.includes('Edge')) {
        const element = document.getElementById(`${this.id}`);
        if (element && element.value.length && this.label) {
          this.floatLabels = true;
          this.inputValue = element.value;
        }
      } else if (navigator.userAgent.includes('Chrome')) {
        const node = this.$refs.input;
        try {
          const nativeMatches = node.matches || node.msMatchesSelector;
          if (nativeMatches.call(node, ':-webkit-autofill') && this.label) {
            this.floatLabels = true;
          }
        } catch (e) {
          noop();
        }
      }
      this.hideLabel = false;
    }, 500, this);

    this.setInputValue(this.value);
  },
  watch: {
    inputValue: {
      handler(newVal) {
        if (newVal !== undefined) {
          this.inputValueHandler(newVal);
        }
      },
      deep: true,
    },
    value(newVal) {
      this.setInputValue(newVal);
    },
  },
  methods: {
    /**
    * Default setInputValue method. Overrides possible in components
    *
    * @param {Array|Object|Number|String} newVal value to be set for internal model
    */
    setInputValue(newVal) {
      if (newVal !== undefined && newVal !== null) {
        this.inputValue = newVal;
      }
    },
    /**
    * Default inputValueHandler method. Overrides possible in components
    *
    * @param {Array|Object|Number|String} newVal value to be set for internal model
    */
    inputValueHandler(newVal) {
      if (newVal === null) {
        this.$emit('input', null);
      } else {
        // for select when value in a key in an object
        const value = typeof newVal === 'object' && Object.hasOwnProperty.call(newVal, 'value') ? newVal.value : newVal;
        this.floatLabels = value.toString().length > 0 && !!this.label;
        this.$emit('input', value);
      }
    },
  },
};
</script>
