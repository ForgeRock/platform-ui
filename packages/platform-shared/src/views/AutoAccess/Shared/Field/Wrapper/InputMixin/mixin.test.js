/**
 * Copyright (c) 2022-2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { shallowMount, flushPromises } from '@vue/test-utils';
import InputMixin from './index';

const TestComponent = {
  render() {},
  mixins: [InputMixin],
};

describe('InputMixin', () => {
  it('InputMixin uid is set', () => {
    const wrapper = shallowMount(TestComponent);
    const expected = `floatingLabelInput${wrapper.vm._uid}`;
    expect(wrapper.vm.$data.id).toBe(expected);
  });

  it('Initial value calls setter method setInputValue', () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        value: 'test',
      },
    });

    expect(wrapper.vm.inputValue).toBe('test');
  });

  it('Initial value defaults to empty string', () => {
    const wrapper = shallowMount(TestComponent);

    expect(wrapper.vm.$data.inputValue).toBe('');
  });

  it('Initial value can be set as String', () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        value: 'test',
      },
    });

    expect(wrapper.vm.$data.inputValue).toBe('test');
  });

  it('Initial value can be set as Array', () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        value: ['test'],
      },
    });

    expect(wrapper.vm.$data.inputValue).toEqual(['test']);
  });

  it('Initial value can be set as Object', () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        value: { test: 'test' },
      },
    });

    expect(wrapper.vm.$data.inputValue).toEqual({ test: 'test' });
  });

  it('Initial value can be set as Number', () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        value: 5,
      },
    });

    expect(wrapper.vm.$data.inputValue).toBe(5);
  });

  it('Update to inputValue triggers handler and event', async () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        label: 'test label',
      },
    });
    const inputValueHandlerSpy = jest.spyOn(wrapper.vm, 'inputValueHandler');

    wrapper.setData({ inputValue: 'test' });
    await flushPromises();

    expect(inputValueHandlerSpy).toBeCalled();
    expect(wrapper.vm.$data.floatLabels).toBeTruthy();
    expect(wrapper.emitted().input[0]).toStrictEqual(['test']);
  });

  it('Update to inputValue triggers handler and event', async () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        label: 'test label',
      },
    });
    const inputValueHandlerSpy = jest.spyOn(wrapper.vm, 'inputValueHandler');

    wrapper.setData({ inputValue: 'test' });
    await flushPromises();

    expect(inputValueHandlerSpy).toBeCalled();
    expect(wrapper.vm.$data.floatLabels).toBeTruthy();
    expect(wrapper.emitted().input[0]).toStrictEqual(['test']);
  });
});

describe('InputMixin floatLabels', () => {
  it('String inputValue change updates floatLabels', async () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        label: 'test label',
      },
    });

    expect(wrapper.vm.$data.floatLabels).toBeFalsy();
    wrapper.setData({ inputValue: 't' });
    await flushPromises();
    expect(wrapper.vm.$data.floatLabels).toBeTruthy();
    wrapper.setData({ inputValue: '' });
    await flushPromises();
    expect(wrapper.vm.$data.floatLabels).toBeFalsy();
  });

  it('Object inputValue change updates floatLabels', async () => {
    const wrapper = shallowMount(TestComponent, {
      props: {
        label: 'test label',
        value: { value: '' },
      },
    });

    expect(wrapper.vm.$data.floatLabels).toBeFalsy();
    wrapper.setData({ inputValue: { value: 't' } });
    await flushPromises();
    expect(wrapper.vm.$data.floatLabels).toBeTruthy();
    wrapper.setData({ inputValue: { value: '' } });
    await flushPromises();
    expect(wrapper.vm.$data.floatLabels).toBeFalsy();
  });
});
