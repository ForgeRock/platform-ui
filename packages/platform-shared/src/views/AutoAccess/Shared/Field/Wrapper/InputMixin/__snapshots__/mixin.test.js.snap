// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputMixin Initial value calls setter method setInputValue: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (3)",
    "
",
    " at <Anonymous",
    "value="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Initial value can be set as Array: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (6)",
    "
",
    " at <Anonymous",
    "value=",
    [
      "test",
    ],
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) "watch" option or vm.$watch on an array value will no longer trigger on array mutation unless the "deep" option is specified. If current usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ WATCH_ARRAY: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/watch.html",
    "
",
    " at <Anonymous",
    "value=",
    [
      "test",
    ],
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Initial value can be set as Number: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (8)",
    "
",
    " at <Anonymous",
    "value=5",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Initial value can be set as Object: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (7)",
    "
",
    " at <Anonymous",
    "value=",
    {
      "test": "test",
    },
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Initial value can be set as String: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (5)",
    "
",
    " at <Anonymous",
    "value="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Initial value defaults to empty string: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (4)",
    "
",
    " at <Anonymous",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin InputMixin uid is set: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (2)",
    "
",
    " at <Anonymous",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Update to inputValue triggers handler and event: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (9)",
    "
",
    " at <Anonymous",
    "label="test label"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin Update to inputValue triggers handler and event: console-warn-snapshot 2`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (10)",
    "
",
    " at <Anonymous",
    "label="test label"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin floatLabels Object inputValue change updates floatLabels: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (12)",
    "
",
    " at <Anonymous",
    "label="test label"",
    "value=",
    {
      "value": "",
    },
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`InputMixin floatLabels String inputValue change updates floatLabels: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (11)",
    "
",
    " at <Anonymous",
    "label="test label"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
